### PR TITLE
[VOL-92] dvol reset in Go

### PR DESCRIPTION
**Pull request by brendan-donegan**
Implements `dvol reset` in Go and makes pass:

- test_reset
- test_reset_HEAD
- test_reset_HEAD_multiple_commits

test_reset_HEAD_hat_multiple_commits does pass if you remove any references to dvol log in it, which will need to be implemented in another branch.

### CODE REVIEW

File: pkg/api/api.go [hunk 1]
```diff
@@ -200,6 +200,18 @@ func (dvol *DvolAPI) Commit(activeVolume, activeBranch, commitMessage string) (s
 	return string(commitId), nil
 }
 
+func (dvol *DvolAPI) ResetActiveVolume(commit string) error {
```
Comment (mdevilliers) [inline pkg/api/api.go]: Do we accept empty  commit strings? I don't know?

File: pkg/cmd/reset.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+			}
+		},
+	}
+	cmd.Flags().BoolVarP(&hardReset, "hard", "", false, "Force removal of newer data (must be set)")
+	return cmd
+}
+
+func resetVolume(cmd *cobra.Command, args []string, out io.Writer) error {
+	dvol := api.NewDvolAPI(basePath)
+	commit := args[0]
```
Comment (mdevilliers) [inline pkg/cmd/reset.go]: Can a commit string be an empty string?

File: pkg/datalayer/datalayer.go [hunk 3]
```diff
@@ -63,11 +64,54 @@ func (dl *DataLayer) RemoveVolume(volumeName string) error {
...
+			return err
+		}
+	} else {
+		commitId = CommitId(commit)
+	}
+	commitPath := dl.commitPath(volumeName, commitId)
+	if _, err := os.Stat(commitPath); err != nil {
+		return err
+	}
+	// TODO: Acquire lock
```
Comment (mdevilliers) [inline pkg/datalayer/datalayer.go]: I assume locking here would make the datalayer go-routine safe? Lets do that!

File: pkg/datalayer/datalayer.go [hunk 4]
```diff
@@ -63,11 +64,54 @@ func (dl *DataLayer) RemoveVolume(volumeName string) error {
...
+	if err := os.RemoveAll(variantPath); err != nil {
+		return err
+	}
+	dl.copyFiles(commitPath, variantPath)
```
Comment (mdevilliers) [inline pkg/datalayer/datalayer.go]: copyFiles returns an error - do we need to check it?

File: pkg/datalayer/datalayer.go [hunk 5]
```diff
@@ -63,11 +64,54 @@ func (dl *DataLayer) RemoveVolume(volumeName string) error {
...
+	dl.destroyNewerCommits(commitId, volumeName, variantName)
```
Comment (mdevilliers) [inline pkg/datalayer/datalayer.go]: destroyNewerCommits returns an error - do we need to check it?

File: pkg/datalayer/datalayer.go (hunk 3 above)
Comment (brendan-donegan) [inline pkg/datalayer/datalayer.go]: Locking is going to come in another branch - it's needed in a couple of places

**Comment by brendan-donegan** [comment]
I did a few changes according to your feedback, maybe you want to review again?

File: pkg/datalayer/datalayer.go [hunk 6]
```diff
@@ -186,3 +234,83 @@ func (dl *DataLayer) WriteCommitsForBranch(volumeName, variantName string, commi
...
+		offset = len(remainder)
+	} else {
+		return "", fmt.Errorf("Malformed commit identifier %s", commit)
+	}
+	// Read the commit database
+	commits, err := dl.ReadCommitsForBranch(volumeName, variantName)
+	return commits[len(commits)-1-offset].Id, err
+}
+
+func indexOfCommit(commitId CommitId, commits []Commit) int {
```
Comment (mdevilliers) [inline pkg/datalayer/datalayer.go]: As discussed change so returns an index (int) and an error of instance NotFound????? 
Sniff in the layer above and return

**Comment by mdevilliers** [comment]
Address and Merge please

**Pull request merged**